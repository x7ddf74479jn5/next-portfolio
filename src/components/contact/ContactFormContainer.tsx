import { yupResolver } from "@hookform/resolvers/yup";
import * as React from "react";
import { SubmitHandler, useForm } from "react-hook-form";
import PrimaryButton from "src/components/common/PrimaryButton";
import ContactFormBody from "src/components/contact/ContactFormBody";
import * as yup from "yup";

import { useModalDispatch } from "../../hooks/useModalDispatch";
import { useModalState } from "../../hooks/useModalState";
import type { FormData } from "./ContactFormBody";

const schema = yup.object().shape({
  name: yup.string().required("お名前が未入力です。"),
  email: yup
    .string()
    .required("メールアドレスが未入力です。")
    .matches(
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
      "メールアドレスの形式に誤りがあります。"
    ),
  category: yup.string(),
  describe: yup.string().required("お問い合わせ内容が未入力です。"),
});
// export type FormData = yup.InferType<typeof schema>;

export const ContactFormContainer = () => {
  const methods = useForm<FormData>();
  // const methods = useForm<FormData>({ mode: "onBlur", resolver: yupResolver(schema) });
  const { handleSubmit, watch } = methods;
  const watcher = watch();

  const { isModalOpen } = useModalState();
  const { openModal, closeModal } = useModalDispatch();
  const [isFetching, setIsFetching] = React.useState(false);

  const isApplyButtonDisabled = React.useMemo(() => {
    if (isFetching) {
      return true;
    }

    return false;
  }, [isFetching]);

  const isConfirmButtonDisabled = React.useMemo(() => {
    if (isModalOpen || isFetching) {
      return true;
    }

    return false;
  }, [isModalOpen, isFetching]);

  const onSubmit = React.useCallback<SubmitHandler<FormData>>(
    (formData: FormData) => {
      openModal("CONTACT", { formData, cancel, apply, isApplyButtonDisabled });
    },
    [openModal]
  );

  const sendMessage = async (data: FormData) => {
    const url = process.env.NEXT_PUBLIC_CONTACT_API_URL;
    if (!url) return;
    return await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
  };

  const apply = async (data: FormData) => {
    setIsFetching(true);
    await sendMessage(data)
      .then(() => {
        alert(
          "お問い合わせありがとうございました。入力いただいたメールアドレスに自動返信メールが届いたことをご確認ください。"
        );
      })
      .catch(() => {
        alert("エラーが発生しました。恐れ入りますが再度送信してください。");
      });
    setIsFetching(false);
    closeModal();
  };

  const cancel = () => {
    closeModal();
  };

  return (
    <form className="p-forms" name="contactForm" onSubmit={handleSubmit(onSubmit)}>
      <ContactFormBody methods={methods} />

      <PrimaryButton id="confirm-btn" type="submit" disabled={isConfirmButtonDisabled}>
        内容を確認する
      </PrimaryButton>
    </form>
  );
};

export default ContactFormContainer;
