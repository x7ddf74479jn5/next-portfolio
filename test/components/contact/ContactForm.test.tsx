import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import ContactFormContainer from "src/components/contact/ContactFormContainer";
import { fireEvent, render, screen } from "test/test-utils";

describe("ContactFormContainer", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<ContactFormContainer />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("first-rendered correctly", () => {
    render(<ContactFormContainer />);
    const formElement = screen.getByRole("form");
    expect(formElement).toBeInTheDocument();

    const nameInput = screen.getByLabelText("お名前 *");
    expect(nameInput).toHaveAttribute("name", "name");
    expect(nameInput).toHaveAttribute("type", "text");
    expect(nameInput).toHaveAttribute("placeholder", "熊猫 鮫");

    const emailInput = screen.getByLabelText("メールアドレス *");
    expect(emailInput).toHaveAttribute("name", "email");
    expect(emailInput).toHaveAttribute("type", "email");
    expect(emailInput).toHaveAttribute("placeholder", "your@example.com");

    const categorySelect = screen.getByLabelText("お問い合わせ種別");
    expect(categorySelect).toHaveAttribute("name", "category");
    expect(categorySelect.children).toHaveLength(5);
    expect(categorySelect.children[0]).toHaveTextContent("");
    expect(categorySelect.children[0]).toHaveAttribute("value", "");
    expect(categorySelect.children[1]).toHaveTextContent("Webサイト制作について");
    expect(categorySelect.children[1]).toHaveAttribute("value", "Webサイト制作について");
    expect(categorySelect.children[2]).toHaveTextContent("Webアプリ開発について");
    expect(categorySelect.children[2]).toHaveAttribute("value", "Webアプリ開発について");
    expect(categorySelect.children[3]).toHaveTextContent("Webサイト模写について");
    expect(categorySelect.children[3]).toHaveAttribute("value", "Webサイト模写について");
    expect(categorySelect.children[4]).toHaveTextContent("その他");
    expect(categorySelect.children[4]).toHaveAttribute("value", "その他");

    const descriptionInput = screen.getByLabelText("お問い合わせ内容 *");
    expect(descriptionInput).toHaveAttribute("name", "description");
    expect(descriptionInput).toHaveAttribute("rows", "10");

    const submitButton = screen.getByRole("button");
    expect(submitButton).toHaveTextContent("内容を確認する");
    expect(submitButton).toHaveAttribute("type", "submit");
    expect(submitButton).toBeDisabled();
  });

  describe("user interaction", () => {
    let nameInput: HTMLElement;
    let emailInput: HTMLElement;
    let descriptionInput: HTMLElement;
    let submitButton: HTMLElement;
    beforeEach(() => {
      render(<ContactFormContainer />);
      nameInput = screen.getByRole("textbox", { name: "お名前 *" });
      emailInput = screen.getByRole("textbox", { name: "メールアドレス *" });
      descriptionInput = screen.getByRole("textbox", { name: "お問い合わせ内容 *" });
      submitButton = screen.getByRole("button");
    });

    it("should display required error when value is invalid", async () => {
      userEvent.click(nameInput);
      expect(nameInput).toHaveFocus();
      userEvent.click(emailInput);
      expect(emailInput).toHaveFocus();
      userEvent.click(descriptionInput);
      expect(descriptionInput).toHaveFocus();
      fireEvent.blur(descriptionInput);

      expect(await screen.findByText("お名前が未入力です。")).toBeInTheDocument();
      expect(await screen.findByText("メールアドレスの形式に誤りがあります。")).toBeInTheDocument();
      expect(await screen.findByText("お問い合わせ内容が未入力です。")).toBeInTheDocument();
      expect(submitButton).toBeDisabled();
    });

    it("should display matching error when email is invalid", async () => {
      userEvent.type(emailInput, "email");
      fireEvent.blur(emailInput);

      expect(emailInput).toHaveValue("email");
      expect(await screen.findByText("メールアドレスの形式に誤りがあります。")).toBeInTheDocument();
    });

    it("should not display error when value is valid", async () => {
      userEvent.type(nameInput, "name");
      userEvent.type(emailInput, "email@test.com");
      userEvent.type(descriptionInput, "description");
      fireEvent.blur(descriptionInput);

      expect(await screen.queryByText("お名前が未入力です。")).not.toBeInTheDocument();
      expect(await screen.queryByText("メールアドレスの形式に誤りがあります。")).not.toBeInTheDocument();
      expect(await screen.queryByText("お問い合わせ内容が未入力です。")).not.toBeInTheDocument();

      expect(nameInput).toHaveValue("name");
      expect(emailInput).toHaveValue("email@test.com");
      expect(descriptionInput).toHaveValue("description");
      await waitFor(() => {
        expect(submitButton).toBeEnabled();
      });
    });

    it("should not display error when value change from invalid to valid", async () => {
      userEvent.click(nameInput);
      expect(nameInput).toHaveFocus();
      userEvent.click(emailInput);
      expect(emailInput).toHaveFocus();
      userEvent.click(descriptionInput);
      expect(descriptionInput).toHaveFocus();
      fireEvent.blur(descriptionInput);

      expect(await screen.findByText("お名前が未入力です。")).toBeInTheDocument();
      expect(await screen.findByText("メールアドレスの形式に誤りがあります。")).toBeInTheDocument();
      expect(await screen.findByText("お問い合わせ内容が未入力です。")).toBeInTheDocument();
      expect(submitButton).toBeDisabled();

      userEvent.type(nameInput, "name");
      userEvent.type(emailInput, "email@test.com");
      userEvent.type(descriptionInput, "description");
      fireEvent.blur(descriptionInput);

      expect(await screen.findByText("お名前が未入力です。")).not.toBeInTheDocument();
      expect(await screen.queryByText("メールアドレスの形式に誤りがあります。")).not.toBeInTheDocument();
      expect(await screen.queryByText("お問い合わせ内容が未入力です。")).not.toBeInTheDocument();

      expect(nameInput).toHaveValue("name");
      expect(emailInput).toHaveValue("email@test.com");
      expect(descriptionInput).toHaveValue("description");
      await waitFor(() => {
        expect(submitButton).toBeEnabled();
      });
    });
  });
});
