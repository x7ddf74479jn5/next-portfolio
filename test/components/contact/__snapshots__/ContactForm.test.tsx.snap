// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactFormContainer matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="p-forms"
    name="contactForm"
  >
    <div
      class="formsInput"
    >
      <p>
        <label
          for="name"
        >
          お名前 *
        </label>
        <input
          id="name"
          name="name"
          placeholder="熊猫 鮫"
          type="text"
        />
      </p>
    </div>
    <div
      class="formsInput"
    >
      <p>
        <label
          for="email"
        >
          メールアドレス *
        </label>
        <input
          id="email"
          name="email"
          placeholder="your@example.com"
          type="email"
        />
      </p>
    </div>
    <div
      class="formsInput"
    >
      <p>
        <label
          for="category"
        >
          お問い合わせ種別
        </label>
        <select
          id="category"
          name="category"
        >
          <option
            value=""
          />
          <option
            value="Webサイト制作について"
          >
            Webサイト制作について
          </option>
          <option
            value="Webアプリ開発について"
          >
            Webアプリ開発について
          </option>
          <option
            value="Webサイト模写について"
          >
            Webサイト模写について
          </option>
          <option
            value="その他"
          >
            その他
          </option>
        </select>
      </p>
    </div>
    <div
      class="formsInput"
    >
      <p>
        <label
          for="description"
        >
          お問い合わせ内容 *
        </label>
        <textarea
          id="description"
          name="description"
          rows="10"
        />
      </p>
    </div>
    <button
      class="btnRound"
      disabled=""
      id="confirm-btn"
      type="submit"
    >
      内容を確認する
    </button>
  </form>
</DocumentFragment>
`;
